core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -307,113 +324,187 @@ private[spark] class ExecutorAllocationManager(
    * number of executors to add next round instead of continuing to double it.
    *
    * @param maxNumExecutorsNeeded the maximum number of executors all currently running or pending
    *                              tasks could fill
+   * @param rpId                  the ResourceProfile id of the executors
    * @return the number of additional executors actually requested.
    */
-  private def addExecutors(maxNumExecutorsNeeded: Int): Int = {
+  private def addExecutors(maxNumExecutorsNeeded: Int, rpId: Int): Int = {
+    val oldNumExecutorsTarget = numExecutorsTargetPerResourceProfileId(rpId)
     // Do not request more executors if it would put our target over the upper bound
-    if (numExecutorsTarget >= maxNumExecutors) {
+    // this is doing a max check per ResourceProfile
+    if (oldNumExecutorsTarget >= maxNumExecutors) {
       logDebug(s"Not adding executors because our current target total " +
-        s"is already $numExecutorsTarget (limit $maxNumExecutors)")
-      numExecutorsToAdd = 1
+        s"is already ${oldNumExecutorsTarget} (limit $maxNumExecutors)")
+      numExecutorsToAddPerResourceProfileId(rpId) = 1
       return 0
     }
-
-    val oldNumExecutorsTarget = numExecutorsTarget
     // There's no point in wasting time ramping up to the number of executors we already have, so
     // make sure our target is at least as much as our current allocation:
-    numExecutorsTarget = math.max(numExecutorsTarget, executorMonitor.executorCount)
+    numExecutorsTargetPerResourceProfileId(rpId) =
+      math.max(numExecutorsTargetPerResourceProfileId(rpId),
+        executorMonitor.executorCountWithResourceProfile(rpId))
+
     // Boost our target with the number to add for this round:
-    numExecutorsTarget += numExecutorsToAdd
+    numExecutorsTargetPerResourceProfileId(rpId) +=
+      numExecutorsToAddPerResourceProfileId.getOrElseUpdate(rpId, 1)
+
     // Ensure that our target doesn't exceed what we need at the present moment:
-    numExecutorsTarget = math.min(numExecutorsTarget, maxNumExecutorsNeeded)
+    numExecutorsTargetPerResourceProfileId(rpId) =
+      math.min(numExecutorsTargetPerResourceProfileId(rpId), maxNumExecutorsNeeded)
+
     // Ensure that our target fits within configured bounds:
-    numExecutorsTarget = math.max(math.min(numExecutorsTarget, maxNumExecutors), minNumExecutors)
+    numExecutorsTargetPerResourceProfileId(rpId) = math.max(
+      math.min(numExecutorsTargetPerResourceProfileId(rpId), maxNumExecutors), minNumExecutors)
 
-    val delta = numExecutorsTarget - oldNumExecutorsTarget
+    val delta = numExecutorsTargetPerResourceProfileId(rpId) - oldNumExecutorsTarget
 
     // If our target has not changed, do not send a message
     // to the cluster manager and reset our exponential growth
     if (delta == 0) {
-      // Check if there is any speculative jobs pending
-      if (listener.pendingTasks == 0 && listener.pendingSpeculativeTasks > 0) {
```

Comment:
note that this code here was extra and not needed anymore, we already add in the pendingspeculative tasks before this is called in the maxNumExecutorsNeeded calculation.